File: play-scala/app/controllers/Application.scala
package controllers

import play.api.Play.current
import play.api.mvc._
import play.api.libs.json._
import scala.concurrent.forkjoin.ThreadLocalRandom
import scala.concurrent._
import java.util.concurrent.{ThreadPoolExecutor, LinkedBlockingQueue, TimeUnit}
import models.World
import utils._
import play.modules.reactivemongo.ReactiveMongoPlugin
import play.modules.reactivemongo.json.collection.JSONCollection
import play.api.libs.concurrent.Execution.Implicits._
import play.core.NamedThreadFactory

object Application extends Controller {

  val database = ReactiveMongoPlugin.db
  def collection: JSONCollection = database.collection[JSONCollection]("world")

  val worldWithoutMongoId = (__ \ "_id").json.prune

  private val MaxQueriesPerRequest = 20
  private val TestDatabaseRows = 10000

  private val partitionCount = current.configuration.getInt("db.default.partitionCount").getOrElse(2)
  private val maxConnections =
    partitionCount * current.configuration.getInt("db.default.maxConnectionsPerPartition").getOrElse(5)
  private val minConnections =
    partitionCount * current.configuration.getInt("db.default.minConnectionsPerPartition").getOrElse(5)

  private val tpe = new ThreadPoolExecutor(minConnections, maxConnections,
    0L, TimeUnit.MILLISECONDS,
    new LinkedBlockingQueue[Runnable](),
    new NamedThreadFactory("dbEc"))
  private val dbEc = ExecutionContext.fromExecutorService(tpe)

  // A predicate for checking our ability to service database requests is determined by ensuring that the request
  // queue doesn't fill up beyond a certain threshold. For convenience we use the max number of connections * the max
  // # of db requests per web request to determine this threshold. It is a rough check as we don't know how many
  // queries we're going to make or what other threads are running in parallel etc. Nevertheless, the check is
  // adequate in order to throttle the acceptance of requests to the size of the pool.
  def isDbAvailable: Boolean = (tpe.getQueue.size() < maxConnections * MaxQueriesPerRequest)


  def json() = Action {
    Ok(Json.obj("message" -> "Hello World!"))
  }

  def db(queries: Int) = PredicatedAction(isDbAvailable, ServiceUnavailable) {
    Action {
      Async {
        val random = ThreadLocalRandom.current()

        val worlds = Future.sequence((for {
          _ <- 1 to queries
        } yield Future(World.findById(random.nextInt(TestDatabaseRows) + 1))(dbEc)
          ).toList)

        worlds map {
          w => Ok(Json.toJson(w))
        }
      }
    }
  }

  def mongodb(queries: Int) = Action {
    import scala.concurrent.ExecutionContext.Implicits.global

    Async {
      val random = ThreadLocalRandom.current()
      val futureList = Future.sequence((for {
        _ <- 1 to queries
      } yield { collection
        .find(Json.obj("id" -> (random.nextInt(TestDatabaseRows) + 1)))
        .cursor[JsObject]
        .toList map {
          l => l.head.transform(worldWithoutMongoId).get
        }
      }))

      futureList.map { worlds =>
        Ok(Json.toJson(worlds))
      }
    }
  }

}

File: play-java-jpa/app/controllers/Application.java
package controllers;

import akka.dispatch.ExecutionContexts;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import models.World;
import play.Play;
import play.core.NamedThreadFactory;
import play.libs.F;
import play.libs.Json;

import play.mvc.Controller;
import play.mvc.Result;
import scala.concurrent.ExecutionContext;
import utils.Predicate;
import utils.Predicated;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.*;

public class Application extends Controller {

    private static final int MAX_QUERIES_PER_REQUEST = 20;
    private static final int TEST_DATABASE_ROWS = 10000;
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private static final int partitionCount = Play.application().configuration().getInt("db.default.partitionCount");
    private static final int maxConnections =
            partitionCount * Play.application().configuration().getInt("db.default.maxConnectionsPerPartition");
    private static final int minConnections =
            partitionCount * Play.application().configuration().getInt("db.default.minConnectionsPerPartition");

    private static final ThreadPoolExecutor tpe = new ThreadPoolExecutor(minConnections, maxConnections,
            0L, TimeUnit.MILLISECONDS,
            new LinkedBlockingQueue<Runnable>(),
            new NamedThreadFactory("dbEc"));
    private static final ExecutionContext dbEc = ExecutionContexts.fromExecutorService(tpe);

    // A predicate for checking our ability to service database requests is determined by ensuring that the request
    // queue doesn't fill up beyond a certain threshold. For convenience we use the max number of connections * the max
    // # of db requests per web request to determine this threshold. It is a rough check as we don't know how many
    // queries we're going to make or what other threads are running in parallel etc. Nevertheless, the check is
    // adequate in order to throttle the acceptance of requests to the size of the pool.
    public static class IsDbAvailable implements Predicate {
        @Override
        public boolean condition() {
            return (tpe.getQueue().size() < maxConnections * MAX_QUERIES_PER_REQUEST);
        }
    }

    public static Result json() {
        final ObjectNode result = OBJECT_MAPPER.createObjectNode();
        result.put("message", "Hello World!");
        return ok(result);
    }


    @Predicated(predicate = IsDbAvailable.class, failed = SERVICE_UNAVAILABLE)
    public static F.Promise<Result> db(final Integer queries) {
        final Random random = ThreadLocalRandom.current();
        final List<F.Promise<? extends World>> promises = new ArrayList<F.Promise<? extends World>>(queries);
        for (int i = 0; i < queries; ++i) {
            final F.Promise<World> p = F.Promise.promise(new F.Function0<World>() {
                @Override
                public World apply() throws Throwable {
                    return World.findById(Long.valueOf(random.nextInt(TEST_DATABASE_ROWS) + 1));
                }
            }, dbEc);
            promises.add(p);
        }
        return F.Promise.sequence(promises).map(new F.Function<List<World>, Result>() {
            @Override
            public Result apply(List<World> worlds) {
                return ok(Json.toJson(worlds));
            }
        });
    }

}
